service/matching/matcher.go

```diff
@@ -286,26 +344,49 @@ forLoop:
 			// in turn dispatched the task to a poller. Make sure we delete the
 			// task from the database
 			task.finish(nil)
+			tm.emitDispatchLatency(task, true)
 			return nil
 		case <-ctx.Done():
 			return ctx.Err()
+		case <-reconsiderFwdTimerC:
+			continue forLoop
 		case <-interruptCh:
 			return errInterrupted
 		}
 	}
 }
 
+func (tm *TaskMatcher) emitDispatchLatency(task *internalTask, forwarded bool) {
+	if task.event.Data.CreateTime == nil {
+		return // should not happen but for safety
+	}
+
+	source := task.source
+	if source == enumsspb.TASK_SOURCE_UNSPECIFIED {
+		// history may not specify the source
+		source = enumsspb.TASK_SOURCE_HISTORY
+	}
+
+	tm.metricsHandler.Timer(metrics.TaskDispatchLatencyPerTaskQueue.GetMetricName()).Record(
+		time.Since(*task.event.Data.CreateTime),
+		metrics.StringTag("source", source.String()),
+		metrics.StringTag("forwarded", strconv.FormatBool(forwarded)),
```

Comment:
I had a crazy idea on metrics here: what if we attached to each task and poll a short coded string. Polls start with the empty string, add "P" each time a poll is forwarded. Tasks start with "H" if direct from history or "D" if it passed through the db, and then add a "T" each time it's forwarded. When we get a match, concatenate the task and poll strings and use that as the metric label. It's one label that includes all the information we want concisely. We can use regexp matches to select different combinations.